### DISCUSSION TITLE
Meshing between PolySurface and Plane

### DISCUSSION THREAD

**Discussion by RauchLukas** [pyvista/pyvista#2548] [Q&A] [answered]
Hey guys, im new within the PyVista community. So my question might be easy to solve and maybe realy dumb. 

**anyway:**

I am trying to calculate the Volume between a arbitrary surface and a plane. 

Therefore i got two approaches in mind. 

____
**1. Extrude the surface, then clip the solid with the plane.**
   
   kind of like in this [example](https://docs.pyvista.org/examples/00-load/terrain-mesh.html#sphx-glr-examples-00-load-terrain-mesh-py)
   This one works fine for the minimal example, but clipping with the plane is not optimal in the in the bigger scope. 

Also, repeating the grid looks like it only works with ```StructuredGrid```. In my case i have to deal with arbitrary meshes.

**2.  Extrude the surface directly to the bounds of the plane.**
   I could not find a straight forward solution to extrude to a plane. 
   -> please let me know if there is some easy solution for that.

  Since i can use the ```poly.project_points_to_plane``` Function, i thought maybe meshing between this two point sets might be the solution, to skip the expensive and error-prone clipping process.

The big question is, how to generate the Mesh/Grid between the vertices, without realy knowing the context of neighboring points.

**Comment by whophil** [comment, marked as answer]
I don't think this is currently covered in PyVista, but the VTK filter `vtkTrimmedExtrusionFilter` sounds like it does what you'd like.

https://vtk.org/doc/nightly/html/classvtkTrimmedExtrusionFilter.html#details

Here is an example
```python
import pyvista as pv
import vtk

mesh = pv.ParametricRandomHills(randomseed=2)
plane = pv.Plane(center=(0, 0, -5), direction=(0, 0, -1), i_size=100, j_size=100)

alg = vtk.vtkTrimmedExtrusionFilter()
alg.SetInputData(0, mesh)
alg.SetInputData(1, plane)
alg.SetCappingStrategy(0) # <-- ensure that the cap is defined by the intersection
alg.SetExtrusionDirection(0, 0, -1.0) # <-- set this with the plane normal
alg.Update()
output = pv.core.filters._get_output(alg)

p = pv.Plotter()
p.add_mesh(plane)
p.add_mesh(mesh)
p.add_mesh(output, color='red')
p.show()

print(output.volume)
```

<img width="413" alt="image" src="https://user-images.githubusercontent.com/2186528/167556837-1a01dda5-5865-464d-a0b6-d60d5104ea08.png">

**Reply by RauchLukas** [reply]
@whophil  Awsom!! that is exaclty what i am looking for. 
Thank for pasting allso this minimal example. 

cheers!!! 
Lukas

**Reply by RauchLukas** [reply]
@whophil, @tkoyama010, 

I saw you are already working on a nativ PyVista implementation of the vtk.vtkTrimmedExtrusionFilter() Method. That is awsom. Thanks for your support, i am really exited. 

At the time i was trying to validate the suggested Method, i am a litte curious, something is not working as expected. 

to keep things simple and avoid errors, i made up an easy example inspired by [Create PolyData](https://docs.pyvista.org/examples/00-load/create-poly.html#sphx-glr-examples-00-load-create-poly-py): 

```python 
import pyvista
import numpy as np
import vtk

p = pyvista.Plotter()
p.show_bounds()

# mesh points
vertices = np.array([[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0.5, 0.5, 1]])

# VOLUME
faces = np.hstack([[4, 0, 1, 2, 3], [3, 0, 1, 4], [3, 1, 2, 4], [3, 2, 3, 4], [3, 3, 0, 4]])  # [square, triangle, triangle, triangle, triangle]
pyramid = pyvista.PolyData(vertices, faces)
print("Volume pyramid = {:1.6} \tis close to {:1.6}" .format(pyramid.volume, 1/3))
assert np.isclose(pyramid.volume, 1/3)

# SURFACE
faces = np.hstack([[3, 0, 1, 4], [3, 1, 2, 4], [3, 2, 3, 4], [3, 3, 0, 4]])  # [triangle, triangle, triangle, triangle]
surf = pyvista.PolyData(vertices, faces)
surf.triangulate(inplace=True)
p.add_mesh(surf, style='wireframe', color='r')

# Cutting Plane
h = 1
trim_surf = pyvista.Plane(center=(surf.center[0], surf.center[1], h), direction=(0,0,h), i_size=2, j_size=2)
p.add_mesh(trim_surf, color='r', opacity=0.3)

alg = vtk.vtkTrimmedExtrusionFilter()
alg.SetInputData(0, surf)
alg.SetInputData(1, trim_surf)
alg.SetCappingStrategy(0) # <-- ensure that the cap is defined by the intersection
alg.SetExtrusionDirection(0, 0, h) # <-- set this with the plane normal
alg.Update()
output = pyvista.core.filters._get_output(alg)


# Checking Results
volume_goal = pyramid.volume + abs(h)
print("Volume extrusion = {:1.6} \tis close to {:1.6}" .format(output.volume, volume_goal))
# assert np.isclose(output.volume, volume_goal)

p.add_mesh(output, color='w', opacity=0.3)

p.show()
```
<img src="https://user-images.githubusercontent.com/57633463/169297769-e3bef9ba-8bbd-4919-ba6b-851ba19d4ea0.png" width="450">

the Test fails, because the volume of the pyramid on top is subtracted instead of added to the volume of the quad. 

for comparison, reversing the direction of the cutting plane to `h = +1`  leads to 

<img src="https://user-images.githubusercontent.com/57633463/169299257-d4706ede-b46d-4eb2-9613-621f52220c42.png" width="450">

Now, the volume of the Pyramid gets added to the volume of the cube. 


To be honest, I am not sure if this is a problem, extrusion `VTK` related or volume integral `PyVista` related, but since the graphical output looks totally right, i guess, something is wrong on the pyvista side.

**Reply by MatthewFlamm** [reply]
This might be a normals issue.  Try your above example with this line instead:

```py
output = pyvista.core.filters._get_output(alg).compute_normals()
```
